File: python/taflow/metrics/entropic_value_at_risk.py
"""Empirical entropic value-at-risk metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import EntropicValueAtRisk as _Native
from ._input import as_metric_series


class EntropicValueAtRisk:
    """Compute empirical positive-loss Entropic Value at Risk.

    For loss observations ``L=-return``, TAFlow minimizes
    ``z * (log(mean(exp(L/z))) + log(1/cutoff))`` over ``z > 0``. The default
    lower-tail probability is ``cutoff=0.05``. This is Riskfolio-Lib 7.3's
    empirical ``EVaR_Hist`` convention: output is a positive loss magnitude
    when the selected tail represents losses, not a signed return quantile.
    The executable oracle independently minimizes that pinned Riskfolio-Lib
    objective with SciPy rather than relying on TAFlow's optimizer.

    TAFlow uses shifted log-sum-exp and deterministic bracketing/bisection of
    the monotone first-order condition. The relative interval tolerance is
    ``1e-12`` with at most 256 iterations. If the infimum occurs as ``z`` tends
    to zero, the exact worst loss is returned. Empty input is ``None``; one
    observation completes warm-up and returns its negated return.

    Exact empirical EVaR cannot be updated from fixed sufficient statistics:
    Rust retains O(n) normalized returns. ``append`` is amortized O(1), bulk
    ``extend`` is one native GIL-free loop, and the first ``compute`` after new
    input is O(n * iterations). Repeated unchanged ``compute`` calls are O(1)
    through a lazy cache. Semantic factories accept simple returns, log
    returns, equity levels, and period P&L with positive initial equity.
    ``nan_policy`` is ``"omit"`` or ``"raise"``; infinities are rejected.
    """

    def __init__(self) -> None:
        """Reject ambiguous construction; use a semantic ``from_*`` factory."""
        raise TypeError(
            "use EntropicValueAtRisk.from_returns/from_equity/from_pnl/"
            "from_log_returns"
        )

    @classmethod
    def _create(
        cls,
        values: Any,
        input_mode: str,
        *,
        cutoff: float = 0.05,
        initial_equity: float | None = None,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "EntropicValueAtRisk":
        state = cls.__new__(cls)
        state._state = _Native(input_mode, float(cutoff), initial_equity, nan_policy)
        return state.extend(values, column=column)

    @classmethod
    def from_returns(
        cls,
        returns: Any,
        *,
        cutoff: float = 0.05,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "EntropicValueAtRisk":
        """Construct from chronological decimal simple returns."""
        return cls._create(
            returns, "returns", cutoff=cutoff, nan_policy=nan_policy, column=column
        )

    @classmethod
    def from_log_returns(
        cls,
        log_returns: Any,
        *,
        cutoff: float = 0.05,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "EntropicValueAtRisk":
        """Construct from chronological log returns converted by Rust."""
        return cls._create(
            log_returns,
            "log_returns",
            cutoff=cutoff,
            nan_policy=nan_policy,
            column=column,
        )

    @classmethod
    def from_equity(
        cls,
        equity: Any,
        *,
        cutoff: float = 0.05,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "EntropicValueAtRisk":
        """Construct from positive equity or adjusted-price levels."""
        return cls._create(
            equity, "equity", cutoff=cutoff, nan_policy=nan_policy, column=column
        )

    @classmethod
    def from_pnl(
        cls,
        pnl: Any,
        *,
        initial_equity: float,
        cutoff: float = 0.05,
        nan_policy: str = "omit",
        column: str | None = None,
    ) -> "EntropicValueAtRisk":
        """Construct from period P&L and required positive initial equity."""
        return cls._create(
            pnl,
            "pnl",
            cutoff=cutoff,
            initial_equity=float(initial_equity),
            nan_policy=nan_policy,
            column=column,
        )

    def append(self, value: float) -> "EntropicValueAtRisk":
        """Append one selected-domain observation and return this metric."""
        self._state.append(float(value))
        return self

    def extend(
        self, values: Any, *, column: str | None = None
    ) -> "EntropicValueAtRisk":
        """Append observations through one native loop and return this metric."""
        self._state.extend(as_metric_series(values, column=column))
        return self

    @property
    def value(self) -> float | None:
        """Return current positive-loss EVaR, or ``None`` when empty."""
        return self._state.value

    def compute(self) -> float | None:
        """Optimize after changed input, otherwise return the O(1) cached scalar."""
        return self._state.compute()

    def reset(self) -> "EntropicValueAtRisk":
        """Clear observations, retain capacity and settings, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the usable normalized-return count delegated to Rust."""
        return len(self._state)


__all__ = ["EntropicValueAtRisk"]

File: crates/taflow-metrics/src/metrics/entropic_value_at_risk.rs
use crate::{MetricError, MetricInputKind, MetricInputState, MetricResult, NanPolicy};

/// Positive-loss empirical Entropic Value at Risk with lazy exact optimization.
#[derive(Debug, Clone)]
pub struct EntropicValueAtRisk {
    input: MetricInputState,
    returns: Vec<f64>,
    cutoff: f64,
    cached_value: Option<f64>,
    dirty: bool,
}

impl EntropicValueAtRisk {
    /// Construct an empty empirical EVaR state.
    pub fn new(cutoff: f64, nan_policy: NanPolicy) -> MetricResult<Self> {
        if !cutoff.is_finite() || cutoff <= 0.0 || cutoff >= 1.0 {
            return Err(MetricError::InvalidParameter {
                name: "cutoff",
                value: cutoff.to_string(),
                reason: "must be finite and strictly between zero and one",
            });
        }

        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            returns: Vec::new(),
            cutoff,
            cached_value: None,
            dirty: false,
        })
    }

    /// Append one chronological observation in amortized O(1) time.
    pub fn append(&mut self, value: f64) -> MetricResult<()> {
        if let Some(simple_return) = self.input.append(value)? {
            self.returns.push(simple_return);
            self.dirty = true;
        }
        Ok(())
    }

    /// Append a chronological slice through one native loop.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<()> {
        self.returns.reserve(values.len());
        for &value in values {
            self.append(value)?;
        }
        Ok(())
    }

    /// Return exact empirical EVaR as a positive-loss measure.
    ///
    /// For losses `L=-return`, this minimizes
    /// `z * (log(mean(exp(L / z))) + log(1 / cutoff))` over `z > 0`.
    /// The monotone first-order condition is bracketed and bisected with a
    /// relative interval tolerance of `1e-12` and at most 256 iterations.
    /// Log-sum-exp evaluation is shifted by the worst loss. If the empirical
    /// optimum is attained only as `z` approaches zero, the exact worst loss
    /// is returned. Exact recomputation requires O(n) retained history and
    /// O(n * iterations) work only after new observations make the cache dirty.
    pub fn value(&mut self) -> Option<f64> {
        if self.returns.is_empty() {
            return None;
        }
        if !self.dirty {
            return self.cached_value;
        }

        let value = self.optimize();
        self.cached_value = Some(value);
        self.dirty = false;
        Some(value)
    }

    /// Return the cached result, optimizing only after input changes.
    pub fn compute(&mut self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior without shrinking retained capacity.
    pub fn reset(&mut self) {
        self.input.reset();
        self.returns.clear();
        self.cached_value = None;
        self.dirty = false;
    }

    /// Return the number of usable normalized returns retained.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable normalized returns are retained.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }

    fn optimize(&self) -> f64 {
        const RELATIVE_TOLERANCE: f64 = 1e-12;
        const MAXIMUM_ITERATIONS: usize = 256;

        let maximum_loss = self
            .returns
            .iter()
            .map(|value| -*value)
            .max_by(f64::total_cmp)
            .expect("nonempty history has a worst loss");
        let minimum_loss = self
            .returns
            .iter()
            .map(|value| -*value)
            .min_by(f64::total_cmp)
            .expect("nonempty history has a best loss");
        let range = maximum_loss - minimum_loss;
        if range == 0.0 {
            return maximum_loss;
        }

        let worst_count = self
            .returns
            .iter()
            .filter(|value| -**value == maximum_loss)
            .count();
        if self.cutoff * self.returns.len() as f64 <= worst_count as f64 {
            return maximum_loss;
        }

        let log_inverse_cutoff = (1.0 / self.cutoff).ln();
        let condition = |scaled_inverse_temperature: f64| {
            self.condition(
                scaled_inverse_temperature,
                maximum_loss,
                range,
                log_inverse_cutoff,
            )
        };

        let mut lower = 0.0;
        let mut upper = 1.0;
        while condition(upper) < 0.0 {
            upper *= 2.0;
        }

        for _ in 0..MAXIMUM_ITERATIONS {
            let midpoint = lower + (upper - lower) * 0.5;
            if condition(midpoint) < 0.0 {
                lower = midpoint;
            } else {
                upper = midpoint;
            }
            if upper - lower <= RELATIVE_TOLERANCE * midpoint.max(1.0) {
                break;
            }
        }

        let optimum = lower + (upper - lower) * 0.5;
        let (log_mean_exponential, _) = self.shifted_log_moment(optimum, maximum_loss, range);
        maximum_loss + range * (log_mean_exponential + log_inverse_cutoff) / optimum
    }

    fn condition(
        &self,
        scaled_inverse_temperature: f64,
        maximum_loss: f64,
        range: f64,
        log_inverse_cutoff: f64,
    ) -> f64 {
        let (log_mean_exponential, weighted_scaled_loss) =
            self.shifted_log_moment(scaled_inverse_temperature, maximum_loss, range);
        scaled_inverse_temperature * weighted_scaled_loss
            - log_mean_exponential
            - log_inverse_cutoff
    }

    fn shifted_log_moment(
        &self,
        scaled_inverse_temperature: f64,
        maximum_loss: f64,
        range: f64,
    ) -> (f64, f64) {
        let mut exponential_sum = 0.0;
        let mut weighted_sum = 0.0;
        for &simple_return in &self.returns {
            let scaled_loss = (-simple_return - maximum_loss) / range;
            let weight = (scaled_inverse_temperature * scaled_loss).exp();
            exponential_sum += weight;
            weighted_sum += weight * scaled_loss;
        }
        (
            (exponential_sum / self.returns.len() as f64).ln(),
            weighted_sum / exponential_sum,
        )
    }
}

crate::impl_return_metric_lifecycle!(EntropicValueAtRisk);
